### PR TITLE
Install required Spine packages from PyPI

### PR DESCRIPTION
**Pull request by ererkka**
Preferred way to install is still using `pip install -r requirements.txt` to get Spine Items as well.

Installation needs to be tested before this is merged, and this required the dependent packages to be in PyPI as well.

### CODE REVIEW

**Comment by ererkka** [comment]
Hooray, all tests pass when installing _everything_ from PyPI!

**Comment by ererkka** [comment]
Hooray, all tests pass when installing _everything_ from PyPI!

**Comment by ererkka** [comment]
Please test by installing Spine Toolbox to a fresh environment with `pip install spinetoolbox’.

**Comment by PekkaSavolainen** [comment]
@ererkka. Made a virtualenv and did `pip install spinetoolbox`. Looks very good but there's this little problem

`dagster 0.10.9 requires pytz<2021.0, but you'll have pytz 2021.1 which is incompatible.`

**Comment by ererkka** [comment]
But requirements.txt is not checked in this workflow (`pip install spinetoolbox`). I got `pytz==2020.5`.

**Comment by PekkaSavolainen** [comment]
Interesting, let me try again

**Comment by PekkaSavolainen** [comment]
Ok, this time `pip install spinetoolbox` did install pytz 2020.5 (compatible version).

**BUT**

`pip install --no-cache-dir spinetoolbox`

installed pytz 2021.1 (incompatible version)
These are the last few lines from running this command.

```
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.

We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

dagster 0.10.9 requires pytz<2021.0, but you'll have pytz 2021.1 which is incompatible.
Successfully installed Jinja2-3.0.1 Mako-1.1.4 MarkupSafe-2.0.1 PyYAML-5.4.1 alembic-1.6.5 attrs-21.2.0 backcall-0.2.0 boto3-1.17.93 botocore-1.20.93 cached-property-1.5.2 certifi-2021.5.30 chardet-4.0.0 click-8.0.1 colorama-0.4.4 coloredlogs-14.0 croniter-1.0.13 cx-Oracle-8.2.1 cycler-0.10.0 dagster-0.10.9 datapackage-1.15.2 decorator-4.4.2 docstring-parser-0.7.1 et-xmlfile-1.1.0 faker-8.7.0 fire-0.4.0 future-0.18.2 gdx2py-2.1.1 gdxcc-8.33.2.post1 grpcio-1.38.0 grpcio-health-checking-1.38.0 humanfriendly-9.2 idna-2.10 ijson-3.1.4 ipykernel-5.5.5 ipython-7.24.1 ipython-genutils-0.2.0 isodate-0.6.0 jedi-0.18.0 jill-0.9.3 jmespath-0.10.0 jsonlines-2.0.0 jsonpointer-2.1 jsonschema-3.2.0 jupyter-client-6.1.12 jupyter-core-4.7.1 kiwisolver-1.3.1 linear-tsv-1.1.0 matplotlib-3.3.0 matplotlib-inline-0.1.2 networkx-2.5.1 numpy-1.20.3 openpyxl-3.0.7 pandas-1.2.4 parso-0.8.2 pendulum-2.1.2 pickleshare-0.7.5 pillow-8.2.0 prompt-toolkit-3.0.18 protobuf-3.17.3 psutil-5.8.0 psycopg2-2.8.6 pygments-2.9.0 pymysql-1.0.2 pyodbc-4.0.30 pyparsing-2.4.7 pyreadline-2.1 pyrsistent-0.17.3 pyside2-5.14.2.3 python-dateutil-2.8.1 python-editor-1.0.4 python-gnupg-0.4.7 pytz-2021.1 pytzdata-2020.1 pywin32-301 pyzmq-22.1.0 qtconsole-5.1.0 qtpy-1.9.0 requests-2.25.1 requests-futures-1.0.0 rfc3986-1.5.0 rx-1.6.1 s3transfer-0.4.2 scipy-1.6.3 semantic-version-2.8.5 shiboken2-5.14.2.3 six-1.16.0 spine-engine-0.10.1rc2 spine-items-0.7.6rc1 spinedb-api-0.12.2rc1 spinetoolbox-0.6.1a2 sqlalchemy-1.3.24 tableschema-1.20.2 tabulate-0.8.9 tabulator-1.53.5 termcolor-1.1.0 text-unidecode-1.3 toposort-1.6 tornado-6.1 tqdm-4.61.1 traitlets-5.0.5 unicodecsv-0.14.1 urllib3-1.26.5 watchdog-2.1.2 wcwidth-0.2.5 wget-3.2 xlrd-2.0.1
WARNING: You are using pip version 20.2.3; however, version 21.1.2 is available.
You should consider upgrading via the 'c:\data\virtualenvs\st_no_cache\scripts\python.exe -m pip install --upgrade pip' command.

(st_no_cache) C:\data\virtualenvs\st_no_cache>
```

Used python 3.8 and made fresh virtualenvs in both cases and they were activated when running the commands.

I also created a third virtualenv, **upgraded pip**, and ran command

`pip install --no-cache-dir spinetoolbox`

This time it installed pytz 2020.5 (compatible version). 

So, user's need to have an up-to-date pip.

**Comment by ererkka** [comment]
I added instructions to update pip to the README. I think we should consider dropping the whole Anaconda/Miniconda conda environment instructions in favour of pipx.

**Comment by PekkaSavolainen** [comment]
> I think we should consider dropping the whole Anaconda/Miniconda/conda environment instructions in favour of pipx (which just went ‘official’ as it’s provided by PyPA).

Sounds like an agenda item for the next dev meeting.

**Comment by ererkka** [comment]
I updated the installation instructions in README, and `pip install spinetoolbox` is now under the Official releases section. Waht do you think?

File: README.md [hunk 1]
```diff
@@ -25,10 +26,26 @@ Guide.
...
+Start the application by running `spinetoolbox` which should now be available on the path.
+Update the application with 
+
+    pipx upgrade spinetoolbox
+
+
+### Without Python 
+
+Binary installer packages of Spine Toolbox can be found 
 [here](https://drive.google.com/drive/folders/1t-AIIwRMl3HiYgka4ex5bCccI2gpbspK)
```
Comment (PekkaSavolainen) [inline README.md:45]: This link should be updated to `https://github.com/Spine-project/Spine-Toolbox/releases`.

File: README.md [hunk 2]
```diff
@@ -25,10 +26,26 @@ Guide.
 
 ## Official releases
 
-Release versions of Spine Toolbox can be found 
+### With Python installed
+
+Spine Toolbox is available on [The Python Package Index (PyPI)](https://pypi.org/project/spinetoolbox/).
+With Python installed in your system, you can get the latest release using [pipx](https://pypa.github.io/pipx/).
+
```
Comment (PekkaSavolainen) [inline README.md:33]: How about adding two small steps here to first install pipx and then update PATH. 

```
pip install pipx
pipx ensurepath
```

Comment (ererkka) [reply README.md:39]: done

File: README.md (hunk 1 above)
Comment (ererkka) [reply README.md]: done

**Comment by manuelma** [comment]
> I updated the installation instructions in README, and `pip install spinetoolbox` is now under the Official releases section. What do you think?

We checked it out in today's dev meeting. We agree that it goes in the right direction, however, we might want to be more explicit about the different options and why somebody would like to pick one or the other. Also we should have a *recommended* way for users who are indecisive. The suggestion is to have three options:

1. Installing from PyPI (recommended): For users who want to stay updated with the latest release and hot fixes.
    a. Install python 3.8 or 3.9 from [link to python 3.8 downloads]
    b. Open terminal and run `pipx` etc.
2. Using the windows installer: For windows users who don't want to install Python and don't need regular updates.
    a. Download the installer from [link]
    b. Double click and follow the instructions
3. Installing from Git (development version): For users who want the bleeding edge.
    a. Download and install miniconda
    b. Etc. (what we already have)

Cc. @PekkaSavolainen

**Comment by PekkaSavolainen** [comment]
Too many words in the section names @manuelma. I liked the titles that @DillonJ proposed.
1. Standard Python Installation
2. Windows binaries
3. Development version

**Comment by manuelma** [comment]
> Too many words in the section names @manuelma

Sounds good @PekkaSavolainen 

I wasn't proposing section names, just outlining the structure.

**Comment by manuelma** [comment]
I almost forgot: I think we also need a section on how to switch from the Git version to the PyPI version. As @PekkaSavolainen commented on slack, there might be slight issues on windows with precedence, and I think we need to help users with that. We have to assume that our users don't know anything about pretty much anything.

I can give it a shot and make the changes I'm suggesting if you're ok with it @ererkka. Otherwise if you prefer to do it yourself to keep the PR 'pure', also fine. Please just let me know.

**Comment by ererkka** [comment]
Please go ahead and edit the branch if you like. I can review then as well. I agree that the three main options are:

1. With standard Python installation and pipx: allows semi-regular updates
    1.1 Installing Python: OS app store/pkg manager or Python.org
    1.2 Install pipx
    1.2 Install Toolbox using `pipx install spinetoolbox`
    1.3 Update: `pipx upgrade ...`
 2. Windows binaries: for the very beginners, won’t/can’t install Python, don’t need updates
 3. Developers: There are actually many levels here.
     a) Just get the bleeding edge: `pipx install git+https://github.com/Spine-project/Spine-Toolbox#egg=spinetoolbox`
     b) Create a virtual environment, clone manually and install with `pip install -e .`
     c) Also clone the other packages (DB API, Items, Engine) and install the whole thing

**Comment by ererkka** [comment]
Developer instructions on how to build distribution and upload to PyPI now in file publishing_to_pypi.rst.

**Comment by ererkka** [comment]
I rewrote the installation instructions. Please review @DillonJ @PekkaSavolainen @manuelma

**Comment by ererkka** [comment]
Some tests are failing, but I think it’s not related to the changes here.

**Comment by manuelma** [comment]
@ererkka just to be sure, can you please provide some review instructions? I imagine we just need to go through the new installation instructions in the README? I think we'd also need a mac os tester (@iasonask)

**Comment by ererkka** [comment]
Oh, yeah, oops:

1.  Test `pipx install spinetoolbox`
2. Read the README installation instructions and tell if those are understandable

**Comment by manuelma** [comment]
We probably want to wait till the branch is stable before testing? I have the impression that you @ererkka are still pushing some commits?

**Comment by ererkka** [comment]
I think it should be ready now. I just realised yesterday, that the stand-alone spine-db-editor command was not yet available after installing from PyPI.

**Comment by PekkaSavolainen** [comment]
I did `pipx install spinetoolbox --force` because a previous installation existed and I was too lazy to remove it. Everything I tested seems to be working very well. Things I tested

1. Execute Julia, Python, Gams, Tools using/not using Jupyter Console
2. Upgraded outdated sqlite db
3. Switched Python and Julia kernel specs
4. Created new Python kernel specs
5. Installed Julia (using Settings->Tools->Install Julia button) and created new kernel specs for this Julia
6. Installed SpineOpt using Settings -> Tools -> Add/Update SpineOpt button

Haven't seen any problems so far.

File: README.md [hunk 3]
```diff
@@ -22,92 +23,92 @@ documentation, original graphics and other material are released under the
...
+or [Python.org](https://www.python.org/downloads/windows/). On OS X, get 
+[the latest release at Python.org](https://www.python.org/downloads/mac-osx/). 
+Most Linux distributions have Python preinstalled, but please check which version 
+you have and update if necessary.
 
-2. Open [Anaconda Prompt](https://docs.anaconda.com/_images/win-anaconda-prompt1.png).
+2. Make sure you have the latest version of pip installed:
 
-3. Create a new environment by typing:
+        python3 -m pip install --upgrade pip
```
Comment (PekkaSavolainen) [inline README.md:52]:
```
C:\>python3 -m pip install --upgrade pip
'python3' is not recognized as an internal or external command,
operable program or batch file.
```

`python3` is a linux thing. How about just `python`?

File: docs/source/publishing_to_pypi.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+
+Uploading
+---------
+
+Before making a real upload, please test using TestPyPI which is a separate 
+instance from the real index server.
+Once a version has been uploaded to PyPI, it cannot be reverted or modified. 
+
+`Register an an account <https://test.pypi.org/account/register/>`_ and ask
```
Comment (PekkaSavolainen) [inline docs/source/publishing_to_pypi.rst:38]: an an?

File: docs/source/publishing_to_pypi.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+::
+
+    $ twine upload --repository testpypi dist/*
+
+See `Using TestPyPI <https://packaging.python.org/guides/using-testpypi/>`_ 
+for more information. To avoid entering your username and password every time,
+see `Keyring support in twine documentation <https://twine.readthedocs.io/en/latest/#keyring-support>`_.
+
+If everything went smoothly, you are ready to upload the the real index.
```
Comment (PekkaSavolainen) [inline docs/source/publishing_to_pypi.rst:52]: the the?

File: README.md [hunk 6]
```diff
@@ -22,92 +23,92 @@ documentation, original graphics and other material are released under the
...
-## Official releases
+Three options for installing Spine toolbox exist: as an application within
+an existing Python installation, using stand-alone installation package and using 
+the source files directly (for developers).
 
-Release versions of Spine Toolbox can be found 
-[here](https://drive.google.com/drive/folders/1t-AIIwRMl3HiYgka4ex5bCccI2gpbspK)
-(only available for 64-bit Windows for now). Download the latest version, install and
-run `spinetoolbox.exe`.
+### With standard Python installation
```
Comment (PekkaSavolainen) [inline README.md:32]: <del>With</del> Standard Python installation

File: README.md (hunk 3 above)
Comment (ererkka) [reply README.md:52]: On my Windows, calling `python3` fires up MS Store to install Python. I thought it would be a quite safe option. I can change that to just `python`. On some systems that might launch some system-default Python 2.7 and that’s not what we want...

File: docs/source/publishing_to_pypi.rst (hunk 5 above)
Comment (ererkka) [reply docs/source/publishing_to_pypi.rst:52]: Will fix. There are actually many places with ‘the the’ in the whole repo :D

File: README.md (hunk 3 above)
Comment (ererkka) [reply README.md:52]: I added a note that, on some systems one might need to call `python3` instead.

Comment (PekkaSavolainen) [reply README.md:52]: Ok, I know what happened. There are these *App execution aliases* in win10 settings. There's a default setting when typing python.exe or python3.exe into command prompt. I turned these settings off some time ago. If `python3 -m pip install --upgrade pip` works with the default windows settings, then this was not actually an issue (sorry).

Comment (PekkaSavolainen) [reply README.md]: It is good now. This is resolved.

**Comment by PekkaSavolainen** [comment]
This is good to go on my behalf. @DillonJ, @manuelma what do you think?

**Comment by manuelma** [comment]
> This is good to go on my behalf. @DillonJ, @manuelma what do you think?

I'm struggling a bit to find time to test this, but if you want my input I can commit to test it this week? Otherwise we can just merge...

**Comment by ererkka** [comment]
I think most valuable would be to read through the README installation instructions and think if they make sense.

**Comment by ererkka** [comment]
I pretty confident that technically this works, since all of the tests pass here and at Spine Items.

File: README.md [hunk 7]
```diff
@@ -22,92 +23,97 @@ documentation, original graphics and other material are released under the
...
-## Development version
+Spine Toolbox is available on [The Python Package Index (PyPI)](https://pypi.org/project/spinetoolbox/).
+This option is the recommended way to get the most recent release version but a
+working Python installation is also required. 
 
-The development version of Spine Toolbox is in the master branch on this repository, and it has all the latest features and bug-fixes.
-Please note that this version has passed automated testing, but has not been completely manually tested.
+---
+**NOTE:** If you already have Spine toolbox installed globally using pip, please first 
+uninstall using `pip uninstall spinetoolbox`.
```
Comment (manuelma) [inline README.md:41]: Don't they need to activate the environment before running `pip uninstall spinetoolbox`?

In my opinion, all instructions need to be hyper-detailed. We should always assume that users don't know the first thing about python.

File: README.md [hunk 8]
```diff
@@ -22,92 +23,97 @@ documentation, original graphics and other material are released under the
...
-## Official releases
+Three options for installing Spine toolbox exist: as an application within
+an existing Python installation, using stand-alone installation package and using 
+the source files directly (for developers).
 
-Release versions of Spine Toolbox can be found 
-[here](https://drive.google.com/drive/folders/1t-AIIwRMl3HiYgka4ex5bCccI2gpbspK)
-(only available for 64-bit Windows for now). Download the latest version, install and
-run `spinetoolbox.exe`.
+### Standard Python installation
```
Comment (manuelma) [inline README.md:33]: I would add (recommended) to the title

**Comment by manuelma** [comment]
> I think most valuable would be to read through the README installation instructions and think if they make sense.

I think I want to edit the README myself a little bit, instead of making plenty of comments here. There are two things that worry me about the current version:

1. Sometimes it talks to somebody who has an idea of Python or is at least good with computers. I have met people who really need to be told *everything*.
2. It provides too many options for some of the steps. I would prefer to provide *one* option so we know what the user is doing.

**Comment by manuelma** [comment]
I'm done. This is good to merge now in my opinion.

I just did some small edits to the README. Basically I tried to make it more authoritative so the user doesn't need to make any decisions. We tell them exactly what to do at each step, with plenty of detail so they feel safe (and hopefully, not too dumb).

I have also removed a couple of things I considered a bit distracting. We can re-add them, but hopefully in some other subsection so they don't interfere with the installation steps. I think installation steps need to be as lean as possible. The things I removed are:
1. Some text about installing python on linux. Linux users are probably happy to figure it out themselves. Also, using the system python is a bit tricky in my experience because one can't modify the main env (but now I'm thinking that pipx might solve all those problems)
2. The option to install with `pipx install git+https://github.com/Spine-project/Spine-Toolbox.git`. I really don't think this is needed, although is cool and good to know.

**Comment by ererkka** [comment]
Looks good on a first glance! :)

File: README.md [hunk 9]
```diff
@@ -25,24 +25,24 @@ Guide.
...
 
 ### Installation from PyPI (recommended)
 
 This is the recommended way to install Spine Toolbox and keep it updated with new releases.
 
 ---
-**NOTE:** If you have previously installed Spine Toolbox using Conda, 
-please follow the steps [here](#uninstalling-a-previous-conda-installation) to uninstall it.
+**NOTE:** If you have previously installed Spine Toolbox in a conda environment, 
+please follow the steps [here](#uninstalling-a-previous-installation-in-a-conda-environment) to uninstall it.
```
Comment (manuelma) [inline README.md:40]: The idea here is that users who installed following the 'old' instructions (currently in master) recognize their situation. I don't know if they're very much aware of what a conda environment is. I think they will remember thought that they installed "using Conda", that's why I used that formulation.
